jgrapht-core/src/main/java/org/jgrapht/alg/CliqueMinimalSeparatorDecomposition.java

```diff
@@ -0,0 +1,452 @@
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+/* -----------------
+ * CliqueMinimalSeparatorDecomposition.java
+ * -----------------
+ * (C) Copyright 2015, by Florian Buenzli and Contributors.
+ *
+ * Original Author:  Florian Buenzli
+ * Contributor(s):   Thomas Tschager
+ *                   Tomas Hruz
+ *                   Philipp Hoppen
+ *
+ * $Id$
+ *
+ * Changes
+ * -------
+ * 06-Feb-2015 : Initial revision (FB);
+ *
+ */
+package org.jgrapht.alg;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+import org.jgrapht.Graph;
+import org.jgrapht.Graphs;
+import org.jgrapht.graph.SimpleGraph;
+import org.jgrapht.graph.Subgraph;
+
+/**
+ * Clique Minimal Separator Decomposition using MCS-M+ and Atoms algorithm as
+ * described in Berry et al. An Introduction to Clique Minimal Separator
+ * Decomposition (2010), DOI:10.3390/a3020197, 
+ * <a href="http://www.mdpi.com/1999-4893/3/2/197">
+ * http://www.mdpi.com/1999-4893/3/2/197</a>
+ * 
+ * <p>
+ * The Clique Minimal Separator (CMS) Decomposition is a procedure that splits a
+ * graph into a set of subgraphs separated by minimal clique separators, adding
+ * the separating clique to each component produced by the separation. At the
+ * end we have a set of atoms. The CMS decomposition is unique and yields the
+ * set of the atoms independent of the order of the decomposition.
+ * 
+ * @author Florian Buenzli <fbuenzli@student.ethz.ch>
+ * @author Thomas Tschager <thomas.tschager@inf.ethz.ch>
+ * @author Tomas Hruz <tomas.hruz@inf.ethz.ch>
+ * @author Philipp Hoppen
+ */
+public class CliqueMinimalSeparatorDecomposition<V, E>
+{
+
+    /** Source graph to operate on */
+    private SimpleGraph<V, E> graph;
+
+    /** Minimal triangulation of graph */
+    private SimpleGraph<V, E> chordalGraph;
+
+    /** Fill edges */
+    private Set<E> fillEdges;
+
+    /** Minimal elimination ordering on the vertices of graph */
+    private LinkedList<V> meo;
+
+    /**
+     * List of all vertices that generate a minimal separator of
+     * <code>chordGraph</code>
+     */
+    private List<V> generators;
+
+    /** Set of clique minimal separators */
+    private Set<Set<V>> separators;
+
+    /** The atoms generated by the decomposition */
+    private Set<Set<V>> atoms;
+
+    /** Map for each separator how many components it produces. */
+    private Map<Set<V>, Integer> fullComponentCount = 
+        new HashMap<Set<V>, Integer>();
+
+    /**
+     * Setup a clique minimal separator decomposition on graph <code>g</code>.
+     * 
+     * @param g The graph to decompose.
+     */
+    public CliqueMinimalSeparatorDecomposition(SimpleGraph<V, E> g)

```

Comment:
This should take an UndirectedGraph (interface) rather than SimpleGraph (concrete class).  This allows the algo to be used against other implementations.  If there are constraints on the graph topology, those can be documented (and asserted if desired).